gdal/frmts/zarr/zarrdataset.cpp

```diff
@@ -0,0 +1,375 @@
+/* ==================================================================== */
+/*                              ZarrDataset                             */
+/* ==================================================================== */
+/************************************************************************/
+
+/************************************************************************/
+/*                            ZarrDataset()                             */
+/************************************************************************/
+
+ZarrDataset::ZarrDataset() :
+    fp(nullptr)
+{
+    std::fill_n(abyHeader, CPL_ARRAYSIZE(abyHeader), static_cast<GByte>(0));
+    xt::init_zsystem();
+}
+
+/************************************************************************/
+/*                           ~ZarrDataset()                             */
+/************************************************************************/
+
+ZarrDataset::~ZarrDataset()
+
+{
+    FlushCache();
+    if( fp != nullptr )
+        CPL_IGNORE_RET_VAL(VSIFCloseL(fp));
+}
+
+/************************************************************************/
+/*                          GetGeoTransform()                           */
+/************************************************************************/
+
+CPLErr ZarrDataset::GetGeoTransform( double *padfTransform )
+
+{
+    return CE_None;
+}
+
+/************************************************************************/
+/*                          GetProjectionRef()                          */
+/************************************************************************/
+
+const char *ZarrDataset::_GetProjectionRef()
+
+{
+    return "";
+}
+
+/************************************************************************/
+/*                              Identify()                              */
+/************************************************************************/
+
+int ZarrDataset::Identify( GDALOpenInfo *poOpenInfo )
+
+{
+    // Check that we are accessing a directory,
+    // and that it contains a ".zarray" file.
+    if( !poOpenInfo->bIsDirectory )
+    {
+        return FALSE;
+    }
+
+    CPLString osMDFilename = CPLFormCIFilename( poOpenInfo->pszFilename,
+                                                ".zarray", nullptr );
+
+    VSIStatBufL sStat;
+    if( VSIStatL( osMDFilename, &sStat ) == 0 )
+        return TRUE;
+
+    return FALSE;
+}
+
+/************************************************************************/
+/*                                Open()                                */
+/************************************************************************/
+
+template <class T>
+GDALDataset* ZarrDataset::get_hierarchy(T& store, GDALOpenInfo* poOpenInfo)
+{
+    xt::xzarr_register_compressor<T, xt::xio_gzip_config>();
+    //xt::xzarr_register_compressor<T, xt::xio_blosc_config>();  // TODO
+
+    // Create a corresponding GDALDataset.
+    ZarrDataset *poDS = new ZarrDataset();
```

Comment:
Can any of the following `zarr_*` calls throw? Then this will leak.
Wrapping it with `std::unique_ptr` and `.release()` ing it at the `return` may be a bit safer way.

Reply:
Yes, they can throw. I wrapped the pointer with `std::unique_ptr`, thanks.

Reply:
> Yes, they can throw.

If xtensor-zarr can throw exceptions, they must be caught by any of the methods of the GDAL driver API (virtual methods, and methods like Open()), since GDAL core doesn't expect exceptions to propagate to it.